File: vipaq/tools/Binacle.ViPaq.Generators/InteropArtifactGenerator.cs
using System.Text.Encodings.Web;
using System.Text.Json;
using Binacle.ViPaq;

namespace Binacle.ViPaq.Generators;

// Serializes each shared interop input with the C# ViPaq library and writes the bytes (base64) to
// artifact-cs.json. The TS tool mirrors this, writing the same shape to artifact-ts.json off the same input.
public sealed class InteropArtifactGenerator : IVectorGenerator
{
	public void Generate()
	{
		var interopDir = RepoLocator.FindInteropDir();
		var inputPath = Path.Combine(interopDir, "input.json");
		var outputPath = Path.Combine(interopDir, "artifact-cs.json");

		var readOptions = new JsonSerializerOptions
		{
			// Keys and properties are both PascalCase, so no case-insensitive matching is needed.
			PropertyNameCaseInsensitive = false,
			ReadCommentHandling = JsonCommentHandling.Skip,
		};

		var inputs = JsonSerializer.Deserialize<InputScenario[]>(File.ReadAllText(inputPath), readOptions)
			?? throw new InvalidOperationException("input.json deserialized to null.");

		var artifacts = new List<Artifact>();
		foreach (var input in inputs)
		{
			var bin = CompactParser.ParseBin(input.Bin);
			var items = CompactParser.ParseItems(input.Items);

			// Serialize as long — long holds the whole interoperable range [0, 2^53 - 1] exactly, and the
			// section width is chosen from the value, not the type, so small values still emit 8-bit sections.
			var bytes = ViPaqSerializer.Serialize<Bin, Item, long>(bin, items);
			var encodingInfo = EncodingInfoHelper.FromByte(bytes[0]).ToLabel();

			artifacts.Add(new Artifact
			{
				Name = input.Name,
				Producer = "csharp",
				EncodingInfo = encodingInfo,
				Base64 = Convert.ToBase64String(bytes),
			});
		}

		var writeOptions = new JsonSerializerOptions
		{
			WriteIndented = true,
			IndentCharacter = '\t',
			IndentSize = 1,
			// Keep base64 '+' and '/' literal instead of \u escapes — the content is plain base64, so this
			// is safe and the committed file stays readable.
			Encoder = JavaScriptEncoder.UnsafeRelaxedJsonEscaping,
		};

		File.WriteAllText(outputPath, JsonSerializer.Serialize(artifacts, writeOptions));

		Console.WriteLine($"Wrote {artifacts.Count} artifact(s) to {outputPath}");
		foreach (var artifact in artifacts)
		{
			Console.WriteLine($"  {artifact.Name} -> {artifact.EncodingInfo} ({artifact.Base64.Length} base64 chars)");
		}
	}
}

File: vipaq/binacle-vipaq/tools/interopArtifactGenerator.ts
import * as fs from "fs";
import * as path from "path";
import ViPaqSerializer from "../src/ViPaqSerializer";
import {encodingInfoFromByte} from "../src/utils";
import {parseBin, parseItems} from "./compactParser";
import {toLabel} from "./encodingInfoLabel";
import {Artifact} from "./Artifact";

// Ports C#: InteropArtifactGenerator. Serializes each shared interop input with the TS ViPaq library and
// writes the bytes (base64) to artifact-ts.json. Mirrors the C# generator off the same input; only the
// compressed bytes differ, because Node's CompressionStream and C#'s GZipStream emit different valid gzip.

interface InputScenario {
	Name: string;
	Bin: string;
	Items: string[];
}

export async function generateInteropArtifact(): Promise<void> {
	const interopDir = path.join(__dirname, "..", "..", "test-vectors", "interop");
	const inputPath = path.join(interopDir, "input.json");
	const outputPath = path.join(interopDir, "artifact-ts.json");

	const inputs: InputScenario[] = JSON.parse(fs.readFileSync(inputPath, "utf8"));

	const artifacts: Artifact[] = [];
	for (const input of inputs) {
		const bin = parseBin(input.Bin);
		const items = parseItems(input.Items);

		const bytes = await ViPaqSerializer.serialize(bin, items);

		artifacts.push(new Artifact(
			input.Name,
			"typescript",
			toLabel(encodingInfoFromByte(bytes[0])),
			Buffer.from(bytes).toString("base64"),
		));
	}

	// Tabs + no trailing newline to match the C# generator's output style.
	fs.writeFileSync(outputPath, JSON.stringify(artifacts, null, "\t"));

	console.log(`Wrote ${artifacts.length} artifact(s) to ${outputPath}`);
	for (const artifact of artifacts) {
		console.log(`  ${artifact.Name} -> ${artifact.EncodingInfo} (${artifact.Base64.length} base64 chars)`);
	}
}
